scripts/elasticsearch_nodes.py

```diff
@@ -0,0 +1,123 @@
+dashboard template, the output is padded with 'blank' entries.
+Example:
+{"nodes": "tVsGH2BSXXYe9sMLEYRcIv,ydG9meZMAGkTnAlPgLljz8,blank,blank"}
+"""
+from itertools import (
+    repeat,
+)
+import json
+import logging
+import re
+import sys
+from time import (
+    sleep,
+)
+
+from botocore.exceptions import (
+    ClientError,
+)
+
+from azul import (
+    config,
+    require,
+)
+from azul.deployment import (
+    aws,
+)
+from azul.logging import (
+    configure_script_logging,
+)
+
+log = logging.getLogger(__name__)
+
+# How many times to retry the request to AWS if an error is encountered.
+max_attempts = 10
+
+
+def main() -> int:
+    i = iter(range(max_attempts))
+    while True:
+        attempt = next(i, None)
+        if attempt is None:
+            log.debug('Error: Max attempts exhausted')
+            return 1
+        else:
+            try:
+                response = aws.cloudwatch.list_metrics(Namespace='AWS/ES',
+                                                       MetricName='CPUUtilization',
+                                                       RecentlyActive='PT3H',
+                                                       Dimensions=[
+                                                           {
+                                                               'Name': 'DomainName',
+                                                               'Value': config.es_domain
+                                                           }
+                                                       ])
+            except ClientError as e:
+                log.debug('Error: ' + e.response['Error']['Code'] + '. Retrying...')
+                sleep(attempt + 1)
+            else:
+                nodes = [
+                    d['Value']
+                    for m in response['Metrics']
+                    for d in m['Dimensions']
+                    if d['Name'] == 'NodeId'
+                ]
+                if len(nodes) > 0:
+                    break
+                else:
+                    log.debug('Error: No node IDs found in response.')
+                    sleep(attempt + 1)
+
+    # The list of nodes IDs is appended with dummy entries to ensure the length
+    # matches the number of placeholders in the template. If our list had less
+    # than the template, Terraform would throw an "Invalid index" error. Also,
+    # the dummy entries cannot be empty strings.
+    placeholder_count = node_placeholder_count()
+    require(placeholder_count > 0,
+            'Error: No node ID placeholders found in template file.')
+    nodes = sorted(nodes)
+    nodes += list(repeat('blank', placeholder_count - len(nodes)))
+
+    # As required by Terraform for an external data source, we output the data
+    # as a map of string values.
+    result = {
+        "nodes": ','.join(nodes)
+    }
+    print(json.dumps(result))
+    return 0
+
+
+def node_placeholder_count():
```

Comment:
Don't roll you own. See if you can use `jmespath` which is already a dependency.

Reply:
JMESPath does not currently support recursive / non-rooted expressions. https://www.github.com/jmespath/jmespath.py/issues/110

Reply:
Why does it need to be?

Reply:
I'd be OK with adding https://github.com/mwilliamson/jq.py as a development dependency.

Reply:
Without traversing the JSON recursively, we would need to specify the path with an Optional Object Identifier (`?`) since not all `widgets` in the JSON have a `metrics` property.

```
(.venv) daniel@Crispin ~/repo/azul1/terraform $ jq '.widgets[].properties.metrics[]? | .[]? | select(type == "string")' cloudwatch_dashboard.template.json
...
"AWS/ES"
"Shards.unassigned"
"DomainName"
"azul-index-prod"
"ClientId"
"542754589326"
"."
"ShardCount"
"ShardRole"
"Primary"
"DomainName"
"azul-index-prod"
"NodeId"
"${local.nodes[0]}"
"ClientId"
"542754589326"
...
```

It does not seem Jmespath supports optional object identifiers:
```
>>> foo = jmespath.search('widgets[].properties.metrics[]? | .[]', body)
Traceback (most recent call last):
...
jmespath.exceptions.LexerError: Bad jmespath expression: Unknown token ?:
widgets[].properties.metrics[]? | .[]
```

`jq.py` does support `?`, however the `|` seems to mean something different to `jq.py`

```
>>> jq.all('.widgets[].properties.metrics[]?[]?', body)
[{'expression': 'nv+ni+nd', 'label': 'notifications', 'id': 'n', 'region': 'us-east-1'}, 'AWS/SQS', ...

>>> jq.all('.widgets[].properties.metrics[]?[]? | (type == "string")', body)
[False, True, True, True, True, ...

```

The recursive search using `jq.py` seems less brittle than specifying a JSON path `.widgets[].properties.metrics[]?[]?`.
```
>>> jq.all('.. | select(type == "string")', body)
['log', "SOURCE '/aws/lambda/azul-indexer-prod-aggregate_retry' ...
```